rfcs/0009-defaults-and-overrides.md

```diff
@@ -0,0 +1,547 @@
+
+**Example E3.** An _override_ policy whose rules _set constraints_ to field values of other policies at a lower level, overriding individual policy values of rules with same identification if those values violate the constraints - _merge granularity problem_
+
+The following example illustrates the possibly unintended consequences of enforcing D/O at [strict levels of granularity](#level-of-granularity-of-compared-policy-rules), and the flip side of the `strategy` field offering a closed set of options (`atomic` vs `merge`).
+
+On one hand, the API is simple and straightforward, and there are no deeper side effects to be concerned about, other than at the two levels provided (entire set or atomic values for each individual policy rule.) On the other hand, this design may require more offline interaction between the actors who manage conflicting policies.
+
+![An override policy whose rules set constraints to field values of other policies at a lower level, overriding individual policy values of rules with same identification if those values violate the constraints - merge granularity problem](0009-defaults-and-overrides-assets/example-e3.png)
+
+### Examples F - Policy rule that deactivates default from higher level
+
+The examples in this section introduce a new field `remove: []string` at the same level as the bare set of policy rules. The value of this field, provided as a list, dictates the default policy rules declared at a higher level to be deactivated ("removed") from the effective policy, specified by name of the policy rules.
+
+**Example F1.** A policy that _deactivates_ a _default_ policy rule set at a higher level
+
+![A policy that deactivates a default policy rule set at a higher level](0009-defaults-and-overrides-assets/example-f1.png)
+
+**Example F2.** A policy that tries to _deactivate_ an _override_ policy rule set a higher level
+
+![A policy that tries to deactivate an override policy rule set a higher level](0009-defaults-and-overrides-assets/example-f2.png)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+WIP
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+WIP
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The following alternatives were considered for the design of the API spec to support D/O:
+- [**API DESIGN 1** - `strategy` field - _RECOMMENDED_](#api-design-1---strategy-field---recommended)
+- [**API DESIGN 2** - `granularity` field](#api-design-2---granularity-field)
+- [**API DESIGN 3** - `when` conditions (at any level of the spec)](#api-design-3---when-conditions-at-any-level-of-the-spec)
+- [**API DESIGN 4** - “path-keys”](#api-design-4---path-keys)
+- [**API DESIGN 5** - JSON patch-like](#api-design-5---json-patch-like)
+
+All the examples in the RFC are based on API DESIGN 1.
+
+## API DESIGN 1 - `strategy` field - _RECOMMENDED_
```

Comment:
This is an interesting design - I'm curious to see how it works in practice.

It _feels_ to me like there are a _lot_ of moving parts here, and it may be difficult for users, particularly new users, to understand how these interactions work.

In Gateway API itself, I've recommended we stick with something more approaching the `merge` strategy, and try not to use structs that can't be compared on a per-field basis, to try to keep the interaction complexity as low as possible. I explicitly did not include a switch to change the mode (as the `strategy` field does) because I'm worried about combinatorial interaction complexity.

To paraphrase Spider-man, with great complexity comes great potential for confusion.

I suspect that it will be easy to build a system of Policies that require a UI or similar to understand the resultant policy, which doesn't seem ideal. But conversely, calculating the resultant policy and putting it on a status somewhere creates a large fanout update problem.

So yes, I'm a little worried about how complex this will end up, but not enough to tell you to stop. :)

Reply:
> IOW, people will do merges one way or another. I personally think it's better if we try to control them (to an extent,) and provide users with some framework to follow.

I agree, this is a good, pragmatic point.

I think that as long as the design thinks carefully through all the discoverability concerns (as in https://gateway-api.sigs.k8s.io/geps/gep-713/#user-discoverability ), then it's worth trying things and seeing what works.